designs/instance-type-settings.md

```diff
@@ -0,0 +1,340 @@
+ 
+Items with a 🔑 symbol call out key points or things of note
+
+## 🔑 Proposed Updated API Surface
+
+**Introduce `v1alpha1/InstanceType` CRD**
+
+```yaml
+apiVersion: karpenter.sh/v1alpha1
+kind: InstanceType
+metadata:
+  name: "c5.large"
+spec:   
+  offerings: # These define the requirement offering possibilities for the instance type
+    - karpenter.sh/capacity-type: spot
+      topology.kubernetes.io/zone: us-west-2b
+      karpenter.sh/price: 0.40
+    - karpenter.sh/capacity-type: spot
+      topology.kubernetes.io/zone: us-west-2c
+  allocatable: # This defines the resources that the instance has for bin-packing, will be merged with known resource quantities
+    hardware-vendor.example/foo: "2"
+    hardware-vendor.example/bar: "20"
+    pods: 10
+    memory: 3865Mi 
+```
+
+
+**Extend `v1alpha5/Provisioner` CRD**
+
+In particular, we will extend the `.spec.kubeletConfiguration` with `kubeReserved` and `evictionThreshold` that will be passed down into the bootstrap scripts.
+
+```yaml
+apiVersion: karpenter.sh/v1alpha5
+kind: Provisioner
+metadata:
+  name: default
+spec:
+  requirements:
+    - key: karpenter.sh/capacity-type
+      operator: In
+      values: ["spot"]
+  limits:
+    resources:
+      cpu: "1000"
+  kubeletConfiguration:
+    kubeReserved:
+      memory: "1Gi"
+      cpu: "300m"
+      ephemeral-storage: "1Gi"
+    systemReserved:
+      memory: "1Gi"
+      cpu: "300m"
+      ephemeral-storage: "1Gi"
+    evictionThreshold:
+      hard:
+        memoryAvailable: "1Gi"
+        imagefsAvailable: "10%"
+        nodefsAvailable: "15%"
+        nodefsInodesFree: "5%"
+      soft:
+        ...
+```
+
+## Instance Type Setting Examples
+
+1. Configure the VM memory allocatable values for `c5.large`
+
+```yaml
+apiVersion: karpenter.sh/v1alpha1
+kind: InstanceType
+metadata:
+   name: "c5.large"
+spec:
+   allocatable:
+      memory: 200Mi
+```
+
+1. Configure custom device resources for custom device resources for `c4.large`
+
+```yaml
+apiVersion: karpenter.sh/v1alpha1
+kind: InstanceType
+metadata:
+  name: "c4.large"
+spec:
+   allocatable:
+      hardware-vendor.example/foo: "2"
+```
+
+1. Specify custom offerings with custom pricing information for `c4.large`
+
+```yaml
+apiVersion: karpenter.sh/v1alpha1
+kind: InstanceType
+metadata:
+  name: "c4.large"
+spec:
+  offerings:
+  - karpenter.sh/capacity-type: on-demand
+    topology.kubernetes.io/zone: us-west-2b
+    karpenter.sh/price: "0.60"
+  - karpenter.sh/capacity-type: spot
+    karpenter.sh/price: "0.40"
+```
+
+## Considerations
+
+### Instance Type Overrides for Other Cloud Providers (Azure/CAPI/etc.)
+
+Because Karpenter is intended to be cloud-agnostic, we need to keep this in mind when adding changes to the API surface. For the initial iteration of the `InstanceType` CRD, we will assume that the instance type will be selected by the `.metadata.name` field which will map to the well-defined Kubernetes node field `node.kubernetes.io/instance-type`.
```

Comment:
```suggestion
Because Karpenter is cloud-agnostic, we need to keep this in mind when adding changes to the API surface. For the initial iteration of the `InstanceType` CRD, we will assume that the instance type will be selected by the `.metadata.name` field which will map to the well-defined Kubernetes node field `node.kubernetes.io/instance-type`.
```